cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -428,6 +428,290 @@ const FunctionDoc make_struct_doc{"Wrap Arrays into a StructArray",
                                    "specified through MakeStructOptions."),
                                   {"*args"},
                                   "MakeStructOptions"};
+template <typename KeyType>
+struct MapArrayLookupFunctor {
+  static Result<int64_t> FindOneMapValueIndex(const Array& keys,
+                                              const Scalar& query_key_scalar,
+                                              const int64_t start, const int64_t end,
+                                              const bool from_back = false) {
+    const auto query_key = UnboxScalar<KeyType>::Unbox(query_key_scalar);
+    int64_t index = 0;
+    int64_t match_idx = -1;
+    ARROW_UNUSED(VisitArrayValuesInline<KeyType>(
+        *keys.data(),
+        [&](decltype(query_key) key) -> Status {
+          if (index < start) {
+            ++index;
+            return Status::OK();
+          } else if (index < end) {
+            if (key == query_key) {
```

Comment:
This indexing check is to make sure the index is between [start, end) as `map_array.offsets()` approach is used, instead we could use something like you previously mentioned (inside `ExecMapArray`):
```cpp
auto map = map_array.value_slice(map_array_idx);
auto keys = checked_cast<const StructArray&>(*map).field(0);
auto items = checked_cast<const StructArray&>(*map).field(1);
```
Does the latter incur some performance penalty? Which approach should be used?

(Ditto below)

Reply:
Hmm, I also see that there's some duplicate logic below for the array case. Instead of having this return a single index, you could organize it as follows:

```
template <typename FoundItem>
Status FindMatchingIndices(const Array& keys, const Scalar& query, const bool from_back, FoundItem callback);

RETURN_NOT_OK(FindMatchingIndices(...field(0), query_key_scalar, from_back, [&](int64_t index) {
  // Scalar case: append to builder or call GetScalar()
  // Array case: append to ItemType builder or List<ItemType> builder
});
```

that way all the logic gets centralized